### DISCUSSION TITLE
signals pytest celery_session_worker

### DISCUSSION THREAD

**Discussion by idahogray** [celery/celery#8688] [General]
Hello everyone, I have a django application that uses celery for background tasks. I want to have automated tests for these tasks, including tests of the `task_failure` signal receivers. A few years ago, when I initially built this system, I was using django's built in unittest test runner and copied some hacky code to run tests. I have since switched to using pytest. I am adding some new celery tasks, with `task_failure` receivers, and I wanted to see if there was a better way to build the tests than the ugly scab I built previously. The [celery documentation](https://docs.celeryq.dev/en/stable/userguide/testing.html) provides information about the built-in pytest features, which looked promising.

Here is a simplified version of what I have built.

``` python
# tasks.py
@shared_task(bind=True)
def my_task(self):
   print("my_task was called")
   raise NotImplementedError

@task_failure.connect(sender=my_task)
def handle_my_task_failure(sender=None, **kwargs):
    print("handle_my_task_failure was called")
```
``` python
# tests.py
from tasks import my_task
import pytest

@pytest.mark.celery(
    task_always_eager=False,
    task_eager_propagates=False,
    result_serializer="json",
    result_backend="file://results",
    pool="thread",
)
@pytest.mark.django_db(transaction=True)
def test_my_task_failure(celery_session_worker):
    with pytest.raises(NotImplementedError):
        result = my_task.delay()
        result.get()
```
When I run the test using `pytest -s tests.py`, pytest prints

``` bash
my_task called
```

No matter what I do, I just can't get the celery signals to work.

Has anyone ever tried to test celery signals using the `celery_session_worker` fixture? My ddg/google searching is failing me. Any clues anyone could provide would be greatly appreciated.

Thanks in advance.

P.S. I also posted this on the [django forums](https://forum.djangoproject.com/t/celery-signals-pytest-celery-session-worker/25761/1)

**Comment by WillGibson** [comment]
@samuele-mattiuzzo have been banging our heads against this issue today. Did you manage to make it work @idahogray?

**Reply by idahogray** [reply]
I run my application in docker containers so I set up tests that run just like they do in production. However, I am still having problems getting them to run successfully in CI for some reason. I haven't solved this 100% yet.

**Reply by WillGibson** [reply]
I'd love to find someone who got it working in pytest with the fixtures.
